plugins/destination/gremlin/client/write.go

```diff
@@ -44,19 +43,21 @@ func (c *Client) WriteTableBatch(ctx context.Context, tableName string, msgs mes
 	pks := table.PrimaryKeys()
 	if len(pks) == 0 {
 		// If no primary keys are defined, use all columns
-		for i := range table.Columns {
-			pks = append(pks, table.Columns[i].Name)
-		}
+		pks = table.Columns.Names()
 	}
-	nonPKs := make(map[string]struct{})
-	for _, c := range table.Columns {
-		if !c.PrimaryKey {
-			nonPKs[c.Name] = struct{}{}
+	valueColumns := make([]string, 0, len(table.Columns)-len(pks))
+	if len(table.Columns)-len(pks) > 0 {
+		// not all columns are a part of "pk", so we need to account for the values
+		for _, col := range table.Columns {
+			if !col.PrimaryKey {
+				valueColumns = append(valueColumns, col.Name)
+			}
 		}
 	}
 
 	g := gremlingo.Traversal_().WithRemote(session).V().HasLabel(tableName)
 	for i := range rows {
+		g = g.V().HasLabel(tableName)
```

Comment:
that's the bull of the fix, got it by looking at https://docs.aws.amazon.com/neptune/latest/userguide/gremlin-efficient-upserts.html#gremlin-upserts-vertices